codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -498,15 +595,70 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
 
             // Encode the entire header block.
             headerBlock = ctx.alloc().buffer();
-            headersEncoder.encodeHeaders(streamId, headers, headerBlock);
+            try {
+                headersEncoder.encodeHeaders(streamId, headers, headerBlock);
```

Comment:
> sincewe can determine the length of the header up front what if we reserved enough room up front for the frame header

I am not sure about it, or better, I thought we could estimate it as well (without using previous accumulated statistics as we do in http 1.1) but I couldn't find an efficient and straightforward method to do it

Reply:
Sorry, I think I shouldn't have been more clear because the term `header` is so overloaded in this context. What I meant was maybe we can skip ahead 9 bytes in `headerBlock` to reserve room for the _frame_ header and then encode the http headers to the same buffer and then go back and write the frame header. That will put the entire first HEADERS frame in the same buffer without having to do another copy regardless of the length of the encoded headers.

Reply:
Let me give it a shot! And it seems a good idea to me!

Reply:
Addressed at aab16822d4f92b52312c2e0c93baf61ff65c37b4

The heuristic here works like that:
- if the padding bytes are less then 64, we can zero efficiently using `writeZero` 
- if we can zero efficiently, let's try skipping the right amount of frame header bytes directly in the header block
- if we have been lucky, we could reuse directly the available space within header block OR (ie no luck) if the bytes to copy from it are below 128, we could allocate a right sized one

I can decide to be more strict and encode the "shifted" headers only if padding == 0, but it seems too strict to me; we can be really be lucky with `256 bytes - HEADERS_FRAME_HEADER_LENGTH = 241` bytes available for headers.